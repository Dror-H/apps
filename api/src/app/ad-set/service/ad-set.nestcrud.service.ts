import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { AdSet } from '../data/ad-set.entity';
import { AdSetRepository } from '../data/ad-set.repository';

@Injectable()
export class AdSetNestCrudService extends TypeOrmCrudService<AdSet> {
  constructor(@Inject(AdSetRepository) repo: AdSetRepository) {
    super(repo);
  }
}
