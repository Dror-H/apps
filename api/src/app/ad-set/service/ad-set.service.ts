import {
  AdSetCreationDTO,
  WorkspaceDTO,
  AdSetDTO,
  ResponseSuccess,
  SupportedProviders,
  TargetingDto,
  CacheTtlSeconds,
} from '@instigo-app/data-transfer-object';
import { ThirdPartyAdSetApiService } from '@instigo-app/third-party-connector';
import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common';
import { Cache } from 'cache-manager';
import { groupBy } from 'lodash';
import { from, Observable } from 'rxjs';
import { map, mergeMap, switchMap } from 'rxjs/operators';
import { AdSet } from '../data/ad-set.entity';
import { AdSetRepository } from '../data/ad-set.repository';

@Injectable()
export class AdSetService {
  @Inject(ThirdPartyAdSetApiService)
  private readonly thirdPartyAdSetApiService: ThirdPartyAdSetApiService;

  @Inject(AdSetRepository)
  private readonly repo: AdSetRepository;

  @Inject(CACHE_MANAGER)
  private readonly cacheManager: Cache;

  async create(adSet: Partial<AdSetCreationDTO>, currentWorkspace: WorkspaceDTO): Promise<AdSetDTO> {
    const { provider, adAccount, campaign } = adSet;
    const { accessToken } = currentWorkspace.owner.getAccessToken({ provider });
    const builtAdSet = await this.thirdPartyAdSetApiService.create({
      accessToken,
      adAccountProviderId: adAccount.providerId,
      provider,
      adSet: adSet,
    });
    builtAdSet.adAccount = adAccount;
    builtAdSet.campaign = campaign;
    return this.repo.save(builtAdSet);
  }

  bulkPatch(options: { adSets: Partial<AdSetDTO>[]; workspace: WorkspaceDTO }): Observable<any> {
    const { adSets, workspace } = options;
    void this.cacheManager.reset();
    return from(Object.entries(groupBy(adSets, 'provider'))).pipe(
      mergeMap(([provider, entries]) => {
        const { accessToken } = workspace.owner.getAccessToken({ provider });
        return this.thirdPartyAdSetApiService.bulkPatch({ provider, accessToken, adSets: entries });
      }),
      switchMap(() => this.repo.save(adSets)),
      map(() => new ResponseSuccess('AdSets successfully updated')),
    );
  }

  async findAdSetTargeting(options: {
    workspace: WorkspaceDTO;
    provider: SupportedProviders;
    adSetId: string;
  }): Promise<TargetingDto> {
    const { adSetId, provider, workspace } = options;
    const cacheKey = `adSetTargeting-${provider}-${adSetId}`;
    let result = await this.cacheManager.get(cacheKey);
    if (result) {
      return result as any;
    }
    const { accessToken } = workspace.owner.getAccessToken({ provider });
    result = await this.thirdPartyAdSetApiService.findAdSetTargeting({
      accessToken,
      adSetId,
      provider,
    });
    await this.cacheManager.set(cacheKey, result, { ttl: CacheTtlSeconds.ONE_HOUR });
    return result as any;
  }
}
