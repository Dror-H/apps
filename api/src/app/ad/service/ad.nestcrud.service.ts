import { Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { Ad } from '../data/ad.entity';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';

@Injectable()
export class AdNestCrudService extends TypeOrmCrudService<Ad> {
  constructor(
    @InjectRepository(Ad)
    public repo: Repository<Ad>,
  ) {
    super(repo);
  }
}
