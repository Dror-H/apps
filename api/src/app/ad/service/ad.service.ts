import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common';
import { AdCreationDTO, AdDTO, AdUpdateDTO, ResponseSuccess, WorkspaceDTO } from '@instigo-app/data-transfer-object';
import { Ad } from '@api/ad/data/ad.entity';
import { ThirdPartyAdApiService } from '@instigo-app/third-party-connector';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Cache } from 'cache-manager';
import { from, Observable } from 'rxjs';
import { groupBy } from 'lodash';
import { map, mergeMap, switchMap } from 'rxjs/operators';

@Injectable()
export class AdService {
  @Inject(ThirdPartyAdApiService)
  private readonly thirdPartyAdApiService: ThirdPartyAdApiService;

  @InjectRepository(Ad)
  private readonly adRepository: Repository<Ad>;

  @Inject(CACHE_MANAGER)
  private readonly cacheManager: Cache;

  async create(ad: Partial<AdCreationDTO<any>>, currentWorkspace: WorkspaceDTO): Promise<AdDTO> {
    const { provider, adAccount } = ad;
    const { accessToken } = currentWorkspace.owner.getAccessToken({ provider });
    const builtAd = await this.thirdPartyAdApiService.create({
      accessToken,
      adAccountProviderId: adAccount.providerId,
      provider,
      ad,
    });
    builtAd.adAccount = adAccount;
    return this.adRepository.save(builtAd);
  }

  async bulkCreate(ads: Partial<AdCreationDTO<any>>[], currentWorkspace: WorkspaceDTO): Promise<Ad[]> {
    const { provider, adAccount, campaign } = ads[0];
    const { accessToken } = currentWorkspace.owner.getAccessToken({ provider });
    const createdAds = await this.thirdPartyAdApiService.createMany({
      accessToken,
      adAccountProviderId: adAccount.providerId,
      provider,
      ads: ads,
    });
    const adSet = ads[0].providerSpecificFields?.adSet;
    const mappedAds = createdAds.map((item) => ({ ...item, adAccount, campaign, adSet }));
    return this.adRepository.save(mappedAds);
  }

  public async updateOne(ad: AdUpdateDTO, currentWorkspace: WorkspaceDTO): Promise<Ad> {
    await this.cacheManager.reset();
    const { accessToken } = currentWorkspace.owner.getAccessToken({ provider: ad.provider });
    await this.thirdPartyAdApiService.updateAd({ provider: ad.provider, accessToken, ad });
    const newAd = {
      id: ad.id,
      name: ad.name,
      objectStorySpec: { ...ad.data, adTemplateType: ad.adTemplateType },
    } as Ad;
    return await this.adRepository.save(newAd);
  }

  bulkPatch(options: { ads: Partial<AdDTO>[]; workspace: WorkspaceDTO }): Observable<any> {
    const { ads, workspace } = options;
    void this.cacheManager.reset();
    return from(Object.entries(groupBy(ads, 'provider'))).pipe(
      mergeMap(([provider, entries]) => {
        const { accessToken } = workspace.owner.getAccessToken({ provider });
        return this.thirdPartyAdApiService.bulkPatch({ provider, accessToken, ads: entries });
      }),
      switchMap(() => this.adRepository.save(ads)),
      map(() => new ResponseSuccess('Ads successfully updated')),
    );
  }
}
