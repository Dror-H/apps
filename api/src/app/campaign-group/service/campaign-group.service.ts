import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { CampaignGroup } from '../data/campaign-group.entity';
import { CampaignGroupRepository } from '../data/campaign-group.repository';

@Injectable()
export class CampaignGroupNestCrudService extends TypeOrmCrudService<CampaignGroup> {
  constructor(@Inject(CampaignGroupRepository) public repo: CampaignGroupRepository) {
    super(repo);
  }
}
