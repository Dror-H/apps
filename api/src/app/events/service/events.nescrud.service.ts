import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { EventsRepository } from '@api/events/data/events.repository';
import { EventsEntity } from '@api/events/data/events.entity';

@Injectable()
export class EventsNestCrudService extends TypeOrmCrudService<EventsEntity> {
  constructor(@Inject(EventsRepository) eventsRepository: EventsRepository) {
    super(eventsRepository);
  }
}
