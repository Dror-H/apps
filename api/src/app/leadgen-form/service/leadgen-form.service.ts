import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { LeadgenForm } from '../data/leadgen-form.entity';
import { LeadgenFormRepository } from '../data/leadgen_form.repository';

@Injectable()
export class LeadgenFormNestCrudService extends TypeOrmCrudService<LeadgenForm> {
  constructor(@Inject(LeadgenFormRepository) public repo: LeadgenFormRepository) {
    super(repo);
  }
}
