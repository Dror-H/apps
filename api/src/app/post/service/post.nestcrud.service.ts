import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { Post } from '../data/post.entity';
import { PostRepository } from '../data/post.repository';

@Injectable()
export class PostNestCrudService extends TypeOrmCrudService<Post> {
  constructor(@Inject(PostRepository) repo: PostRepository) {
    super(repo);
  }
}
