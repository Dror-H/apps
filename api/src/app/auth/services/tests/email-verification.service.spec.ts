import { Test } from '@nestjs/testing';
import * as faker from 'faker';
import { EmailVerificationService } from '../email-verification.service';

describe('EmailVerificationService test suite', () => {
  let emailService: EmailVerificationService;
  let userRepository;
  let authService;
  let configService;
  let mailerService;
  let prettyLinkService: any;

  beforeEach(async () => {
    const module = await Test.createTestingModule({
      providers: [
        EmailVerificationService,
        { provide: 'UserRepository', useValue: {} },
        { provide: 'ConfigService', useValue: {} },
        { provide: 'MailerService', useValue: {} },
        { provide: 'AuthService', useValue: {} },
        { provide: 'PrettyLinkService', useValue: {} },
      ],
    }).compile();

    emailService = module.get<EmailVerificationService>(EmailVerificationService);
    userRepository = module.get('UserRepository');
    authService = module.get('AuthService');
    configService = module.get('ConfigService');
    mailerService = module.get('MailerService');
    prettyLinkService = module.get('PrettyLinkService');
  });

  it('should be defined', () => {
    expect(emailService).toBeDefined();
  });

  it('sendRegisterMail - should return success message', async () => {
    // Arrange
    const email = faker.datatype.uuid();
    const expectedResult = `Verification email successfully sent: ${email}`;
    const user = { email };
    const token = faker.datatype.uuid();

    userRepository.findByEmail = jest.fn().mockResolvedValue(user);
    authService.createJwtToken = jest.fn().mockResolvedValue({ token });
    emailService.buildMailerOptions = jest.fn().mockReturnValue({});
    mailerService.sendMail = jest.fn().mockResolvedValue({});
    // Act
    const result = await emailService.sendRegisterMail({ email });
    // Assert
    expect(mailerService.sendMail).toHaveBeenCalled();
    expect(result).toEqual({ message: expectedResult });
  });

  it('sendRegisterMail - should return error message', async () => {
    // Arrange
    const email = faker.datatype.uuid();
    const expectedResult = `Could not send verification email: ${email}`;
    const user = { email };
    const token = faker.datatype.uuid();

    userRepository.findByEmail = jest.fn().mockResolvedValue(user);
    authService.createJwtToken = jest.fn().mockResolvedValue({ token });
    emailService.buildMailerOptions = jest.fn().mockReturnValue({});
    mailerService.sendMail = jest.fn().mockRejectedValue({});
    // Act and Assert
    await expect(emailService.sendRegisterMail({ email })).rejects.toThrow(expectedResult);
  });

  it('confirmEmail - should update user and return it', async () => {
    // Arrange
    const token = faker.datatype.uuid();
    const email = faker.datatype.uuid();
    const user = { id: faker.datatype.uuid(), email, emailVerification: false };
    const expectedResult = { ...user, emailVerification: true };
    authService.verifyToken = jest.fn().mockResolvedValue({ user: { email } });
    userRepository.findByEmail = jest.fn().mockResolvedValue(user);
    userRepository.save = jest.fn().mockResolvedValue(expectedResult);
    // Act
    const result = await emailService.confirmEmail({ token });
    // Assert
    expect(userRepository.save).toBeCalledWith(expectedResult);
    expect(result).toEqual(expectedResult);
  });

  it('confirmEmail - should throw error when when email is already verified', async () => {
    // Arrange
    const error = 'Email already verified!';
    const token = faker.datatype.uuid();
    const email = faker.datatype.uuid();
    const userToken = { email };
    const user = { email, emailVerification: true };
    authService.verifyToken = jest.fn().mockResolvedValue({ user: userToken });
    userRepository.findByEmail = jest.fn().mockResolvedValue(user);
    // Act and Assert
    await expect(emailService.confirmEmail({ token })).rejects.toThrow(error);
  });

  it('buildMailerOptions - should create mailer options', async () => {
    // Arrange
    const email = 'lupu60@gmail.com';
    const token = 'dasndsadknsan32321321';
    const username = 'lupu60';

    configService.get = jest.fn().mockReturnValueOnce('http://instigo.api.io');
    prettyLinkService.emailVerificationShortUrl = jest.fn().mockReturnValueOnce('https://shorturl.at/bhiY4');
    // Act
    const result = await emailService.buildMailerOptions({ userEmail: email, token, name: username });
    // Assert
    expect(result).toMatchSnapshot();
  });
});
