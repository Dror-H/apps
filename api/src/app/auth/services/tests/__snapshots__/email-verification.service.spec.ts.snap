// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailVerificationService test suite buildMailerOptions - should create mailer options 1`] = `
Object {
  "context": Object {
    "action_url": "https://shorturl.at/bhiY4",
    "name": "lupu60",
  },
  "subject": "Register email verification - Instigo",
  "template": "verification-email",
  "to": "lupu60@gmail.com",
}
`;
